core/iwasm/libraries/thread-mgr/thread_manager.c

```diff
@@ -625,16 +625,68 @@ wasm_cluster_set_debug_inst(WASMCluster *cluster, WASMDebugInstance *inst)
 
 #endif /* end of WASM_ENABLE_DEBUG_INTERP */
 
+/* Check whether the exec_env is in one of all clusters, the caller
+   should add lock to the cluster list before calling us */
+static bool
+clusters_have_exec_env(WASMExecEnv *exec_env)
+{
+    WASMCluster *cluster = bh_list_first_elem(cluster_list);
+    WASMExecEnv *node;
+
+    while (cluster) {
+        node = bh_list_first_elem(&cluster->exec_env_list);
+
+        while (node) {
+            if (node == exec_env) {
+                return true;
+            }
+            node = bh_list_elem_next(node);
+        }
+
+        cluster = bh_list_elem_next(cluster);
```

Comment:
should we search the whole cluster list?

Seems we should only search the current cluster, otherwise we may get some thread of other wasm instance?

Reply:
But how to search the current cluster? Seems that we can only get it from exec_env->cluster, right? But if the exec_env has been removed from cluster and destroyed, accessing exec_env will cause crash.

How about, when a cluster is found that it contains the exec_env, then we check whether exec_env->cluster == cluster again? To ensure that the exec_env belongs to that cluster?

Reply:
> How about, when a cluster is found that it contains the exec_env, then we check whether exec_env->cluster == cluster again? To ensure that the exec_env belongs to that cluster?

Sounds reasonable, and I think an `assert` is enough because this should never happen, when it really happens, that means the whole system is in an unstable state